smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ArtifactMetadata.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.codegen.core;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import software.amazon.smithy.model.node.FromNode;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeMapper;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.ToNode;
+
+/**
+ * Class that defines information about an artifact that was created from a model for the trace file.
+ *
+ * <p>ArtifactMetadata contains metadata that
+ * can be instantiated from a {@link software.amazon.smithy.model.node.Node} or written
+ * to a {@link software.amazon.smithy.model.node.Node}. ArtifactMetadata contains the
+ * following information:
+ * <ul>
+ *     <li>id - The identifier of the artifact. For example, Java packages should use the Maven artifact ID. </li>
+ *     <li>version - The version of the artifact (for example, the AWS SDK release number).</li>
+ *     <li>timestamp - The RFC 3339 date and time that the artifact was created.</li>
+ *     <li>type - The type of artifact. For code generation, this is the programming language. </li>
+ *     <li>typeVersion (Optional) - The artifact type version, if relevant. For example, when defining
+ *     trace files for Java source code, the typeVersion would be the minimum supported JDK version.
+ *     Different artifacts may have different output based on the version targets (for example the ability
+ *     to use more features in a newer version of a language). </li>
+ *     <li>homepage (Optional) - The homepage URL of the artifact.</li>
+ * </ul>
+ */
+public class ArtifactMetadata implements ToNode, FromNode, ValidateRequirements {
+    public final String idText = "id";
```

Comment:
nit: These constants should be uppercase with underscores:
```suggestion
    public final String ID_TEXT = "id";
```